### ISSUE TITLE
speed question

### ISSUE THREAD

**Issue by Whatsetsthisend** [Dobiasd/frugally-deep#226]
Why is my same model 10 times slower in frugally-deep than when I use the same CPU instead of GPU in Python TF 2.0

**Comment by Whatsetsthisend** [comment]
i use Visuat studio c++ Release x64 ，Will it be slower than g++? I see in your readme file that the speed drop is usually less than double, and I'm down ten times

**Comment by Dobiasd** [comment]
> Why is my same model 10 times slower in frugally-deep than when I use the same CPU instead of GPU in Python TF 2.0

Depending on the model type, maybe because TF is using multiple CPU cores for one single prediction. Frugally-deep only supports using multiple CPU cores for multiple predictions: https://github.com/Dobiasd/frugally-deep/blob/master/FAQ.md#does-frugally-deep-support-multiple-cpus

> i use Visuat studio c++ Release x64 ，Will it be slower than g++?

Usually no. The Optimizer of MSVC++ is quite good. Just make sure you also allow it to get the most of your CPU architecture by setting the right optimization flags.

> I see in your readme file that the speed drop is usually less than double, and I'm down ten times

The README does not give any comparison between MSVC and g++.

In Summary:

- Make sure to have set the maximum optimization level (and your CPU target architecture) for the MSVC compiler. Also don't attach a debugger, even in release mode.
- If you want to compare with TensorFlow, not just disable the GPU, but also restrict your TF process to using just one CPU (core).

**Comment by Whatsetsthisend** [comment]
What I see is a comparison of the speed of TF and frugally-deep in the readme

**Comment by Dobiasd** [comment]
Yes, and

> If you want to compare Frugally-deep with TensorFlow, not just disable the GPU, but also restrict the TF process to using just one CPU (core).

That's how the comparison in the README was done. (It's stated there too.) :slightly_smiling_face:

Or instead of `model::predict` use `m̀odel::predict_multi` (with multiple input data sets) in frugally-deep, to also utilize multiple cores.

**Comment by Whatsetsthisend** [comment]
thank，and i test in my python code on just one cpu it takes 0.25s,but in frugally-deep it takes 2.5s.

**Comment by Dobiasd** [comment]
Can you upload the used `.h5` model file, so I can verify the performance difference locally?

**Comment by Whatsetsthisend** [comment]
yes，i upload it yestorday,But I don't mind uploading it again，and i will upload the result in my python code and the datasets.I have opened the optimization of MSVC to complete optimization (02)


[data.zip](https://github.com/Dobiasd/frugally-deep/files/4687118/data.zip)

**Comment by Dobiasd** [comment]
Ah ok, so it's the same model.

I just compared locally and can confirm a significant performance difference.

In TensorFlow (with `CUDA_VISIBLE_DEVICES='' taskset --cpu-list 1`) I get a time of roughly `0.6 s` for one forward pass.

With frugally-deep I get roughly `2.9 s` (`g++ -O3`) or `1.8 s` (`g++ -O3 -march=native`).

([`Dockerfile` used for that](https://gist.github.com/Dobiasd/9a076bfee3c71c30052e468726c726a6))

So it's not a factor of 10 on my machine, but still, more than I'd like to see. I'll investigate and get back to you with the results.

**Comment by Whatsetsthisend** [comment]
oh,now i open AVX2 in my MSVC,it  roughly takes 2s，it progress 0.5s than before

**Comment by Dobiasd** [comment]
The model architecture looks pretty standard:

![kAY3pOB](https://i.imgur.com/kAY3pOB.png)

And the shapes also have no surprises:
- `"input_shapes": [[384, 384, 1]]`
- `"output_shapes": [[48, 48, 1]]`

I'm on it. :slightly_smiling_face:
Let me know if you find something interesting during profiling.

**Comment by Dobiasd** [comment]
I've repeated the test outside of Docker and the results are similar:
- ~ `0.55 s` with TensorFlow (`CUDA_VISIBLE_DEVICES='' taskset --cpu-list 1`)
- ~ `1.67 s` with frugally-deep (`g++ -O3 -march=native`)

Also, I tested with an older version of frugally-deep (`v0.13.0-p0` instead of `v0.14.0-p0`), still similar results.

Profiling also does not show any big surprises:

![haXUHQC](https://i.imgur.com/haXUHQC.png)

So, currently, I still have no idea where the too-big difference on that model comes from.

**Comment by Dobiasd** [comment]
Maybe TensorFlow does some clever optimizations on the Conv2D-BatchNormalization-Conv2D-BatchNormalization operations, like kernel fusion or similar, and thus saved time. That's something frugally-deep currently can not do.

So right now I can't help you get the same performance in frugally-deep as in TensorFlow on that model. Sorry.

**Comment by Whatsetsthisend** [comment]
oh,thank you for your help,I found that there is basically a significant difference from tf in the convolutional layer. Perhaps the precision conversion of the convolution operation can greatly improve the performance, because there seems to be such an operation on tf, I have not tried it, yesterday I feel tired world, today I will start to adjust the source code, if I have any progress I will pull the request.

**Comment by Dobiasd** [comment]
I just released a [new version `v0.14.1-p0`](https://github.com/Dobiasd/frugally-deep/releases/tag/v0.14.1-p0) that includes performance improvements of convolution (`Conv2D`) on large input tensors, while also reducing memory usage.

On my machine, this version improves the time spent on a forward pass using your model to `1.13 s` (from previously `1.67 s`). I hope that helps. :slightly_smiling_face:

**Comment by Dobiasd** [comment]
FYI: I'm working on another improvement of the conv2d performance in [pull request 229](https://github.com/Dobiasd/frugally-deep/pull/229). It improves the runtime of our model on my machine further, i.e., to `0.89 s`.
